crates/polars-stream/src/executor/mod.rs

```diff
@@ -0,0 +1,340 @@
+            } else {
+                self.global_low_prio_task_queue.push(task);
+            }
+            self.park_group.unpark_one();
+        }
+    }
+
+    fn try_steal_task<R: Rng>(&self, thread: usize, rng: &mut R) -> Option<ReadyTask> {
+        // Try to get a global task.
+        loop {
+            match self.global_high_prio_task_queue.steal() {
+                Steal::Empty => break,
+                Steal::Success(task) => return Some(task),
+                Steal::Retry => std::hint::spin_loop(),
+            }
+        }
+
+        loop {
+            match self.global_low_prio_task_queue.steal() {
+                Steal::Empty => break,
+                Steal::Success(task) => return Some(task),
+                Steal::Retry => std::hint::spin_loop(),
+            }
+        }
+
+        // Try to steal tasks.
+        let ttl = &self.thread_task_lists[thread];
+        for _ in 0..4 {
+            let mut retry = true;
+            while retry {
+                retry = false;
+
+                for idx in random_permutation(self.thread_task_lists.len() as u32, rng) {
+                    let foreign_ttl = &self.thread_task_lists[idx as usize];
+                    match foreign_ttl
+                        .high_prio_tasks_stealer
+                        .steal_batch_and_pop(&ttl.high_prio_tasks)
+                    {
+                        Steal::Empty => {},
+                        Steal::Success(task) => return Some(task),
+                        Steal::Retry => retry = true,
+                    }
+                }
+
+                std::hint::spin_loop()
+            }
+        }
+
+        None
+    }
+
+    fn runner(&self, thread: usize) {
+        TLS_THREAD_ID.set(thread);
+
+        let mut rng = SmallRng::from_rng(&mut rand::thread_rng()).unwrap();
+        let mut worker = self.park_group.new_worker();
+
+        loop {
+            let ttl = &self.thread_task_lists[thread];
+            let task = (|| {
+                // Try to get a task from LIFO slot.
+                if let Some(task) = unsafe { (*ttl.local_slot.get()).take() } {
+                    return Some(task);
+                }
+
+                // Try to get a local high-priority task.
+                if let Some(task) = ttl.high_prio_tasks.pop() {
+                    return Some(task);
+                }
+
+                // Try to steal a task.
+                if let Some(task) = self.try_steal_task(thread, &mut rng) {
+                    return Some(task);
+                }
+
+                // Prepare to park, then try one more steal attempt.
+                let park = worker.prepare_park();
+                if let Some(task) = self.try_steal_task(thread, &mut rng) {
+                    return Some(task);
+                }
+                park.park();
+                None
+            })();
+
+            if let Some(task) = task {
+                worker.recruit_next();
```

Comment:
What is the semantics of this one? We have found a task and can execute some work, but before we do we wake another thread? Or does it give its `recruiting` job to another thread? 

Where can I find the logic of a recruiter unparking someone?

Reply:
The logic is found in `park_group.rs`. The idea is that there is always at most one 'recruiting' thread that is responsible for unparking threads. So even if there are 100 calls to `ParkGroup::unpark_one()`, it will only unpark one thread and make it a recruiter. Then once that one thread finds a task it will unpark the next one and pass along the recruiter tag. That's what the `recruit_next` function is for - all the docs is on the `ParkGroup`.

The idea is that when a bunch of work becomes available you don't constantly generate a lot of calls to the slow `condvar` in the worker threads, and it also prevents waking up a bunch of threads that then all start contesting trying to steal available tasks.